libs/infinity_emb/infinity_emb/transformer/vision/utils.py

```diff
@@ -3,24 +3,72 @@
 
 from infinity_emb._optional_imports import CHECK_PIL, CHECK_REQUESTS
 from infinity_emb.primitives import ImageCorruption, ImageSingle
+from typing import Union, List, Sequence
+import concurrent.futures
 
 if CHECK_PIL.is_available:
     from PIL import Image  # type: ignore
 if CHECK_REQUESTS.is_available:
     import requests  # type: ignore
 
+MAX_WORKERS = 10
 
-def resolve_images(image_urls: list[str]) -> list[ImageSingle]:
-    """Resolve images from URLs."""
-    # TODO: improve parallel requests, safety, error handling
-    CHECK_REQUESTS.mark_required()
-    CHECK_PIL.mark_required()
 
+def resolve_from_img_obj(img_obj):
+    """Resolve an image from a PIL.Image.Image Object."""
     try:
-        downloaded = [requests.get(url, stream=True).raw for url in image_urls]
+        return ImageSingle(image=img_obj)
     except Exception as e:
-        raise ImageCorruption(f"Error downloading images: {e}")
+        raise ImageCorruption(f"error opening image from obj: {e}")
+
+
+def resolve_from_img_url(img_url):
+    """Resolve an image from an URL."""
     try:
-        return [ImageSingle(image=Image.open(content)) for content in downloaded]
+        downloaded_img = requests.get(img_url, stream=True).raw
     except Exception as e:
-        raise ImageCorruption(f"Error opening images: {e}")
+        raise ImageCorruption(f"error downloading image from url: {e}")
+
+    try:
+        return ImageSingle(image=Image.open(downloaded_img))
+    except Exception as e:
+        raise ImageCorruption(f"error opening image from url: {e}")
+
+
+def resolve_image(img: Union[str, Image.Image]) -> ImageSingle:
+    """Resolve a single image."""
+    if isinstance(img, Image.Image):
+        return resolve_from_img_obj(img)
+    elif isinstance(img, str):
+        return resolve_from_img_url(img)
+    else:
+        raise ValueError(
+            f"Invalid image type: {img} is neither str nor PIL.Image.Image object"
+        )
+
+
+def resolve_images(images: Sequence[Union[str, Image.Image]]) -> List[ImageSingle]:
+    """Resolve images from URLs or PIL.Image.Image Objects using multithreading."""
+    # TODO: improve parallel requests, safety, error handling
+    CHECK_REQUESTS.mark_required()
+    CHECK_PIL.mark_required()
+
+    resolved_imgs = []
+    exceptions = []
+    with concurrent.futures.ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
```

Comment:
Generally an acceptable idea to use ThreadPoolExecutor here, but let's not use it for now.
Mostly concern is that there are many other threads in Infinity already, making this introduction complicated.